Implement StorageManager.estimate()

szewai opened the pull request:
#### a793d8b68a12b37482e8734a578926405f7edc45
<pre>
Implement StorageManager.estimate()
<a href="https://bugs.webkit.org/show_bug.cgi?id=248918">https://bugs.webkit.org/show_bug.cgi?id=248918</a>
rdar://problem/103380414

Reviewed by Chris Dumez.

Spec: <a href="https://storage.spec.whatwg.org/#dom-storagemanager-estimate">https://storage.spec.whatwg.org/#dom-storagemanager-estimate</a>

* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-indexeddb.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-indexeddb.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-parallel.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-parallel.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-application-cache.https.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-caches.https.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-caches.https.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-indexeddb.https.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-indexeddb.https.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-service-workers.https.tentative.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details.https.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details.https.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/idlharness.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/idlharness.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/opaque-origin.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/storagemanager-estimate.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/storagemanager-estimate.https.any.worker-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/storage/DummyStorageProvider.h:
* Source/WebCore/Modules/storage/StorageConnection.h:
* Source/WebCore/Modules/storage/StorageEstimate.h: Copied from Source/WebCore/Modules/storage/StorageManager.idl.
(WebCore::StorageEstimate::isolatedCopy const):
* Source/WebCore/Modules/storage/StorageManager.cpp:
(WebCore::StorageManager::estimate):
* Source/WebCore/Modules/storage/StorageManager.h:
* Source/WebCore/Modules/storage/StorageManager.idl:
* Source/WebCore/Modules/storage/WorkerStorageConnection.cpp:
(WebCore::WorkerStorageConnection::getEstimate):
(WebCore::WorkerStorageConnection::didGetEstimate):
* Source/WebCore/Modules/storage/WorkerStorageConnection.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::estimate):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::estimate):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/NetworkProcess/storage/QuotaManager.cpp:
(WebKit::QuotaManager::usage):
* Source/WebKit/NetworkProcess/storage/QuotaManager.h:
(WebKit::QuotaManager::quota const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebStorageConnection.cpp:
(WebKit::WebStorageConnection::getEstimate):
* Source/WebKit/WebProcess/WebCoreSupport/WebStorageConnection.h:

Canonical link: <a href="https://commits.webkit.org/258610@main">https://commits.webkit.org/258610@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71f326a9ec84e95c4216866906cec01239a88ac0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102497 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11631 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35531 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111764 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106466 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12607 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2532 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/109480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108277 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/94767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24395 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/92744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5089 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25816 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/89066 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2772 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5250 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2263 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29605 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11268 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45309 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/91991 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5919 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6982 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20609 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/storage/StorageEstimate.h:
```diff
@@ -0,0 +1,74 @@
+    unsigned long long quota { 0 };
+
+    StorageEstimate isolatedCopy() const &;
+    StorageEstimate isolatedCopy() &&;
```
Are we anticipating to have more fields?
In any case, we do not not need & and && variants.

szewai replied:
We don't anticipate more fields now -- there was some discussion about showing breakdown of usage details, but I am not sure if it will be made into spec.
Will make the change.

youennf commented on Source/WebCore/Modules/storage/StorageManager.cpp:
```diff
@@ -110,6 +111,22 @@ void StorageManager::persist(DOMPromiseDeferred<IDLBoolean>&& promise)
+            return promise.reject(result.releaseException());
+
+        auto [usage, quota] = result.releaseReturnValue();
+        promise.resolve(StorageEstimate { usage, quota });
```
It seems result is ExceptionOr<StorageEstimate>.
result.releaseReturnValue() is a StorageEstimate, why do we need to recreate it?
Can we directly use promise.settle(WTFMove(result))?

szewai replied:
oops, will change (I recreated the object as I made the result a `std::pair` in previous patch, but I changed it to `StorageEstimate` later)

youennf commented on Source/WebKit/WebProcess/WebCoreSupport/WebStorageConnection.cpp:
```diff
@@ -51,6 +52,16 @@ void WebStorageConnection::persist(const WebCore::ClientOrigin& origin, StorageC
 
+void WebStorageConnection::getEstimate(WebCore::ClientOrigin&& origin, StorageConnection::GetEstimateCallback&& completionHandler)
+{
+    connection().sendWithAsyncReply(Messages::NetworkStorageManager::Estimate(WTFMove(origin)), [completionHandler = WTFMove(completionHandler)](auto result) mutable {
```
WTFMove for origin not needed.

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
```diff
@@ -658,6 +658,13 @@ void OriginStorageManager::setPersisted(bool value)
+{
+    ASSERT(!RunLoop::isMain());
+
+    return WebCore::StorageEstimate { quotaManager().usage(), quotaManager().quota() };
```
This is a potential privacy issue. We need to fuzzy enough the usage() so that the website cannot know the exact size of no-cors opaque responses.
The old cache storage code path is making sure of that and we need to apply similar principles.

szewai replied:
The usage is fuzzy as we kept the cache storage padding behavior in new implementation.
Or do you mean we also want fuzz in other types of usage?

cdumez commented on Source/WebCore/Modules/storage/StorageEstimate.h:
```diff
@@ -0,0 +1,63 @@
+namespace WebCore {
+
+struct StorageEstimate {
+    unsigned long long usage { 0 };
```
Should use uint64_t instead of `unsigned long long`. `unsigned long long` in Web IDL maps to `uint64_t` in C++.

cdumez commented on Source/WebCore/Modules/storage/StorageEstimate.h:
```diff
@@ -0,0 +1,63 @@
+
+template<class Decoder> std::optional<StorageEstimate> StorageEstimate::decode(Decoder& decoder)
+{
+    std::optional<unsigned long long> usage;
```
Needs fixing too.

cdumez commented on Source/WebCore/Modules/storage/StorageEstimate.h:
```diff
@@ -0,0 +1,63 @@
+    template<class Decoder> static std::optional<StorageEstimate> decode(Decoder&);
+};
+
+template<class Encoder> void StorageEstimate::encode(Encoder& encoder) const
```
Ideally, new code would do serialization in the modern format, using WebCoreArgumentCoder.serialization.in.

cdumez commented on Source/WebCore/Modules/storage/StorageManager.cpp:
```diff
@@ -110,6 +111,21 @@ void StorageManager::persist(DOMPromiseDeferred<IDLBoolean>&& promise)
+
+    auto connectionInfo = connectionInfoOrException.releaseReturnValue();
+    connectionInfo.connection.getEstimate(WTFMove(connectionInfo.origin), [promise = WTFMove(promise)](auto&& result) mutable {
+        if (result.hasException())
```
Can we use promise.settle() instead of an if check with reject and resolve?

cdumez commented on Source/WebCore/Modules/storage/StorageManager.idl:
```diff
@@ -32,6 +32,12 @@
 ] interface StorageManager {
     Promise<boolean> persisted();
     [Exposed=Window] Promise<boolean> persist();
+    [EnabledBySetting=StorageAPIEstimateEnabled]Promise<StorageEstimate> estimate();
```
Missing space after `]`.

cdumez commented on Source/WebKit/NetworkProcess/storage/QuotaManager.cpp:
```diff
@@ -44,6 +44,14 @@ QuotaManager::QuotaManager(uint64_t quota, GetUsageFunction&& getUsageFunction,
+uint64_t QuotaManager::usage()
+{
+    if (!m_usage)
+        m_usage = m_getUsageFunction();
```
I don't quite yet why it is OK to cache m_usage? Where do we invalidate m_usage if usage changes?

szewai replied:
We will update m_usage during quota check (when computed usage is bigger than quota, we will fetch the latest usage).
~~But it seems we relied on m_usage is null to decide if we are doing the first quota check and update initial quota based on that, so I guess we'd better not cache it here. Will change.~~
Actually I think we could update m_quota when setting m_usage.

szewai committed:
Implement StorageManager.estimate()
https://bugs.webkit.org/show_bug.cgi?id=248918
rdar://problem/103380414

Reviewed by Chris Dumez.

Spec: https://storage.spec.whatwg.org/#dom-storagemanager-estimate

* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-indexeddb.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-indexeddb.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-parallel.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-parallel.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-application-cache.https.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-caches.https.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-caches.https.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-indexeddb.https.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-indexeddb.https.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details-service-workers.https.tentative.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details.https.tentative.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/estimate-usage-details.https.tentative.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/idlharness.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/idlharness.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/opaque-origin.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/storagemanager-estimate.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/storagemanager-estimate.https.any.worker-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/storage/DummyStorageProvider.h:
* Source/WebCore/Modules/storage/StorageConnection.h:
* Source/WebCore/Modules/storage/StorageEstimate.h: Copied from Source/WebCore/Modules/storage/StorageManager.idl.
(WebCore::StorageEstimate::isolatedCopy const):
* Source/WebCore/Modules/storage/StorageManager.cpp:
(WebCore::StorageManager::estimate):
* Source/WebCore/Modules/storage/StorageManager.h:
* Source/WebCore/Modules/storage/StorageManager.idl:
* Source/WebCore/Modules/storage/WorkerStorageConnection.cpp:
(WebCore::WorkerStorageConnection::getEstimate):
(WebCore::WorkerStorageConnection::didGetEstimate):
* Source/WebCore/Modules/storage/WorkerStorageConnection.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::estimate):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::estimate):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/NetworkProcess/storage/QuotaManager.cpp:
(WebKit::QuotaManager::usage):
* Source/WebKit/NetworkProcess/storage/QuotaManager.h:
(WebKit::QuotaManager::quota const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebStorageConnection.cpp:
(WebKit::WebStorageConnection::getEstimate):
* Source/WebKit/WebProcess/WebCoreSupport/WebStorageConnection.h:

Canonical link: https://commits.webkit.org/258610@main
